### PR TITLE
8326615: C1/C2 don't handle allocation failure properly during initialization (RuntimeStub::new_runtime_stub fatal crash)

### PR DESCRIPTION
**Pull request by dafedafe**
# Issue

The test `compiler/startup/StartupOutput.java` fails intermittently due to a crash after correctly printing the error `Initial size of CodeCache is too small` (the test limits the code cache using `-XX:InitialCodeCacheSize=1024K -XX:ReservedCodeCacheSize=1200k`).
The appearance of the issue is very dependent on thread scheduling. The original report happens during C1 initialization but C2 initialization is affected as well.

# Causes

There is one occurrence during C1 initialization  and one during C2 initialization where a call to `RuntimeStub::new_runtime_stub` can fail fatally if there is not enough space left.
For C1: `Compiler::init_c1_runtime` -> `Runtime1::initialize` -> `Runtime1::generate_blob_for` -> `Runtime1::generate_blob` -> `RuntimeStub::new_runtime_stub`.
For C2: `C2Compiler::initialize` -> `OptoRuntime::generate` -> `OptoRuntime::generate_stub` -> `Compile::Compile` -> `Compile::Code_Gen` -> `PhaseOutput::install` -> `PhaseOutput::install_stub` -> `RuntimeStub::new_runtime_stub`.

# Solution

Instead of avoiding the crash it makes more sense to increase the minimum code cache size by adding the size of the minimal code cache needed for C1 and C2 to `CodeCacheMinimumUseSpace`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326615](https://bugs.openjdk.org/browse/JDK-8326615): C1/C2 don't handle allocation failure properly during initialization (RuntimeStub::new_runtime_stub fatal crash) (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19280/head:pull/19280` \
`$ git checkout pull/19280`

Update a local copy of the PR: \
`$ git checkout pull/19280` \
`$ git pull https://git.openjdk.org/jdk.git pull/19280/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19280`

View PR using the GUI difftool: \
`$ git pr show -t 19280`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19280.diff">https://git.openjdk.org/jdk/pull/19280.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19280#issuecomment-2121873610)

### CODE REVIEW

File: src/hotspot/share/c1/c1_Compiler.cpp [hunk 1]
```diff
@@ -48,16 +48,17 @@
 Compiler::Compiler() : AbstractCompiler(compiler_c1) {
 }
 
-void Compiler::init_c1_runtime() {
+bool Compiler::init_c1_runtime() {
   BufferBlob* buffer_blob = CompilerThread::current()->get_buffer_blob();
-  Runtime1::initialize(buffer_blob);
+  if (!Runtime1::initialize(buffer_blob)) return false;
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_Compiler.cpp:53]:
```suggestion
  if (!Runtime1::initialize(buffer_blob)) {
    return false;
  }
```

File: src/hotspot/share/c1/c1_Runtime1.cpp [hunk 2]
```diff
@@ -259,13 +259,16 @@ void Runtime1::generate_blob_for(BufferBlob* buffer_blob, StubID id) {
...
 }
 
-void Runtime1::initialize(BufferBlob* blob) {
+bool Runtime1::initialize(BufferBlob* blob) {
   // platform-dependent initialization
   initialize_pd();
   // generate stubs
-  for (int id = 0; id < number_of_ids; id++) generate_blob_for(blob, (StubID)id);
+  for (int id = 0; id < number_of_ids; id++) {
+    if (!generate_blob_for(blob, (StubID) id)) return false;
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_Runtime1.cpp:270]:
```suggestion
    if (!generate_blob_for(blob, (StubID) id)) {
      return false;
    }
```

**Review by TobiHartmann — APPROVED** [review]
This is not a regression in JDK 23, right? Could you please adjust the affects versions in JIRA accordingly?

Looks good to me otherwise.

**Comment by dafedafe** [comment]
> This is not a regression in JDK 23, right? Could you please adjust the affects versions in JIRA accordingly?

It is not. Fixing the version.
Thanks a lot for reviewing @TobiHartmann.

**Review by TobiHartmann — APPROVED** [review]
Thanks!

File: src/hotspot/share/c1/c1_Runtime1.cpp [hunk 3]
```diff
@@ -280,6 +285,7 @@ void Runtime1::initialize(BufferBlob* blob) {
 #endif
   BarrierSetC1* bs = BarrierSet::barrier_set()->barrier_set_c1();
   bs->generate_c1_runtime_stubs(blob);
```
Comment (dean-long) [inline src/hotspot/share/c1/c1_Runtime1.cpp:287]: Don't we need to handle failures in generate_c1_runtime_stubs?  With the assert removed, I think we'll get a nullptr crash.

Comment (dafedafe) [reply src/hotspot/share/c1/c1_Runtime1.cpp]: Yep, you're right. Actually that call could potentially fail too.
I've added code to handle that case as well.
Thanks @dean-long!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

**Comment by dafedafe** [comment]
> Looks good to me otherwise.

Thanks for the review @TobiHartmann!

**Comment by dean-long** [comment]
This looks OK, but isn't it a lot of changes just to get this test to pass?  Aren't all of these allocation failures ultimately fatal?  Is there a simpler way to handle this problem?

**Comment by dafedafe** [comment]
> This looks OK, but isn't it a lot of changes just to get this test to pass? Aren't all of these allocation failures ultimately fatal? Is there a simpler way to handle this problem?

It seems a bit much indeed but I think there is potentially always the possibility of not failing but only disabling the compiler. @dean-long do you think the VM would anyway fail later on?

**Comment by dean-long** [comment]
It may not fail, but if it can't create a C1 or C2 compiler, then that's bad, and we might argue that this kind of failure should be similar to a failure during JVM startup.  In fact, I have been thinking that there are reasons why we might want these compiler stubs to be created earlier in startup, when we are still single-threaded.  That would get rid of any races, and make a failure fatal.  It would also allow us to allow us to put these stubs, if they are effectively execute-only because they don't need to be patched, into a special JIT region, avoiding the MAP_JIT overhead on macos-aarch64 (see JDK-8331978).

**Comment by vnkozlov** [comment]
The error message is clear:
```
#  Internal Error (codeBlob.cpp:429), pid=155190, tid=155378
#  fatal error: Initial size of CodeCache is too small
```
I don't think we need to hide it.

**Comment by dafedafe** [comment]
> I suggest instead adjust CodeCache minimal size check when JIT compilers are enabled in `CompilerConfig::check_args_consistency()`. We check `CodeCacheMinimumUseSpace` value their which is 400Kb only. It could be enough to run with `-Xint` (Interpreter only). But it is definitely not enough to run with JIT compilers.

@vnkozlov thanks for your insights!
I've tried to figure out an estimate of a bare minimum code cache size for interpreter/C1/C2/both for a couple of architectures (basically by running `java -version`):
|                | Int  | C1 only | C2 only | C1/C2 tiered |
|----------------|------|---------|---------|--------------|
| linux-x64      | 680k | 4m      | 900k    | 2.2m         |
| macosx-aarch64 | 580k | 3m      | 800k    | 1.5m         |
| linux-aarch64  | 520k | 3.8m    | 800k    | 1.5m         |

(interesting that C2 needs less code cache than C1 but I guess it could also be due to the fact that its thread might actually exit before it properly starts)

800k for the interpreter alone and 5 times as much (4m) when C1 or C2 are running might be good minima. What do you think?

**Comment by vnkozlov** [comment]
@dafedafe  Did you run on avx512 x64 machine? We generate a lot of avx512 specific stubs for C2's intrinsics. We need to count them.

How you got 4Mb for C1 only? I got:
```
java -XX:+PrintCodeCache -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -version
...
CodeCache: size=49152Kb used=1084Kb max_used=1084Kb free=48067Kb
```

**Comment by vnkozlov** [comment]
It was on old AVX2 machine.

**Comment by dafedafe** [comment]
> @dafedafe Did you run on avx512 x64 machine? We generate a lot of avx512 specific stubs for C2's intrinsics. We need to count them.

Yes, I did run it on an avx512 machine. For C2 only I get:
```
CodeCache: size=49152Kb used=928Kb max_used=949Kb free=48223Kb
```

> How you got 4Mb for C1 only? I got:

Strange. I get:

```
java -XX:+PrintCodeCache -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -version
...
CodeCache: size=49152Kb used=3922Kb max_used=3922Kb free=45229Kb
```

**Comment by vnkozlov** [comment]
This is run on big (48 cores) avx512 box with C1 only:
```
$ ./jdk_git/build/product/images/jdk/bin/java -XX:+PrintCodeCache -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:+PrintCompilation -version
32    1       1       java.lang.Object::<init> (1 bytes)
33    2       1       java.lang.String::hashCode (60 bytes)
33    3     n 0       jdk.internal.misc.Unsafe::getReferenceVolatile (native)   
34    4     n 0       jdk.internal.vm.Continuation::enterSpecial (native)   (static)
34    5     n 0       jdk.internal.vm.Continuation::doYield (native)   (static)
java version "24-internal" 2025-03-18
Java(TM) SE Runtime Environment (build 24-internal-2024-07-03-1645504...)
35    6       1       java.lang.String::coder (15 bytes)
Java HotSpot(TM) 64-Bit Server VM (build 24-internal-2024-07-03-1645504..., mixed mode, emulated-client, sharing)
CodeCache: size=49152Kb used=1126Kb max_used=1126Kb free=48025Kb
 bounds [0x00007f4fec83f000, 0x00007f4fecaaf000, 0x00007f4fef83f000]
 total_blobs=279, nmethods=6, adapters=201, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0
```

**Comment by vnkozlov** [comment]
C2 only (no C2 compilations):
```
$ ./jdk_git/build/product/images/jdk/bin/java -XX:+PrintCodeCache -XX:-TieredCompilation -XX:+PrintCompilation -version
32    1     n       jdk.internal.misc.Unsafe::getReferenceVolatile (native)   
33    2     n       jdk.internal.vm.Continuation::enterSpecial (native)   (static)
33    3     n       jdk.internal.vm.Continuation::doYield (native)   (static)
java version "24-internal" 2025-03-18
Java(TM) SE Runtime Environment (build 24-internal-2024-07-03-1645504.vkozlov...)
Java HotSpot(TM) 64-Bit Server VM (build 24-internal-2024-07-03-1645504.vkozlov..., mixed mode, sharing)
CodeCache: size=49152Kb used=612Kb max_used=626Kb free=48539Kb
 bounds [0x00007fb817f6b000, 0x00007fb8181db000, 0x00007fb81af6b000]
 total_blobs=250, nmethods=3, adapters=201, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0
```

**Comment by dafedafe** [comment]
Max used code cache size running a simple HelloWorld:
|                | Int   | C1 only | C2 only | C1/C2 tiered |
|----------------|-------|---------|---------|--------------|
| linux-x64      | 785Kb | 5400Kb  | 1521Kb  | 6219Kb       |
| macosx-aarch64 | 673Kb | 3210Kb  | 1357Kb  | 4701Kb       |
| linux-aarch64  | 671Kb | 6142Kb  | 1462Kb  | 6890Kb       |

**Comment by vnkozlov** [comment]
@dafedafe, can you show full output (including version lines) of `java -XX:+PrintCodeCache -XX:TieredStopAtLevel=1 -version` on linux-x64?
I want to see what consumes CodeCache in your experiments.

**Comment by dafedafe** [comment]
@vnkozlov here it is (sorry for the delay, I updated the repo as well but the results are similar):
```
build/linux-x64/jdk/bin/java -XX:+PrintCodeCache -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -version
java version "24-internal" 2025-03-18
Java(TM) SE Runtime Environment (build 24-internal-2024-07-23-1502438.dfenacci...)
Java HotSpot(TM) 64-Bit Server VM (build 24-internal-2024-07-23-1502438.dfenacci..., mixed mode, emulated-client)
CodeCache: size=49152Kb used=3486Kb max_used=3486Kb free=45665Kb
 bounds [0x000070510d000000, 0x000070510d370000, 0x0000705110000000]
 total_blobs=916, nmethods=556, adapters=284, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0
```

There seem to be many more methods that get compiled with C1 than with C2 when just running `-version`:

- `java -XX:+PrintCodeCache -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:+PrintCompilation -version`:[C1.log](https://github.com/user-attachments/files/16350954/C1.log)

- `java -XX:+PrintCodeCache -XX:-TieredCompilation -XX:+PrintCompilation -version`: [C2.log](https://github.com/user-attachments/files/16350951/C2.log)

**Comment by vnkozlov** [comment]
> build/linux-x64/jdk/bin/java

You are running **Exploded** JDK which does not have CDS and most likely compiling modules loading code.
You should run `build/linux-x64/images/jdk/bin/java` instead which is used in production.
I assume you are building jdk with `make jdk-image`.

**Comment by dafedafe** [comment]
> You should run build/linux-x64/images/jdk/bin/java instead which is used in production.

Oh, I see! Thanks @vnkozlov!
Here are the updated values:

### `-version`
|                | Int   | C1 only | C2 only | C1/C2 tiered |
|----------------|-------|---------|---------|--------------|
| linux-x64      | 515Kb | 1125Kb  | 604Kb   | 1200Kb       |
| macosx-aarch64 | 479Kb | 1080Kb  | 542Kb   | 1135Kb       |
| linux-aarch64  | 478Kb | 1085Kb  | 542Kb   | 1166Kb       |

### `HelloWorld.java`
|                | Int   | C1 only | C2 only | C1/C2 tiered |
|----------------|-------|---------|---------|--------------|
| linux-x64      | 720Kb | 6956Kb  | 1289Kb  | 5771Kb       |
| macosx-aarch64 | 609Kb | 2760Kb  | 1138Kb  | 3768Kb       |
| linux-aarch64  | 608Kb | 6833Kb  | 1155Kb  | 5827Kb       |

I was still puzzled by C1 values being higher than C2 ones but noticed that most of the difference is made by the _C1 temporary CodeBuffer_(s) (563Kb each on `linux-x64`).

**Comment by vnkozlov** [comment]
C1 has much lower compilation threshold so it is expected it has more nmethods. And yes, it has big static temporary buffer - more cores you have, more C1 compiler threads and more these temp buffers. C2 allocates temporary buffer only during compilation and freeing it after that.

**Comment by vnkozlov** [comment]
Based on that we need to scale CodeCacheMinimumUseSpace based on number of C1 compiler threads to take into account buffer size: `NMethodSizeLimit` (or more precise `Compiler::code_buffer_size()`). And may be similar for C2 even so its buffer is not permanent. Size for C2 is difficult to determine because it is calculated based on compilation information. But may be we can use the same as for C1 as rough estimation.

File: src/hotspot/share/compiler/compilerDefinitions.inline.hpp [hunk 4]
```diff
@@ -25,6 +25,8 @@
 #ifndef SHARE_COMPILER_COMPILERDEFINITIONS_INLINE_HPP
 #define SHARE_COMPILER_COMPILERDEFINITIONS_INLINE_HPP
 
+#include "c1/c1_Compiler.hpp"
+#include "opto/c2compiler.hpp"
```
Comment (vnkozlov) [inline src/hotspot/share/compiler/compilerDefinitions.inline.hpp:29]: These should be also under  COMPILER*_PRESENT macros.

Comment (dafedafe) [reply src/hotspot/share/compiler/compilerDefinitions.inline.hpp]: Right! Fixed. Thanks @vnkozlov.

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — APPROVED** [review]
That looks good to me too.

**Comment by dafedafe** [comment]
Thanks @vnkozlov @TobiHartmann for your reviews!

**Pull request closed without merging**